open_spiel/games/othello.cc

```diff
@@ -0,0 +1,344 @@
+    col = std::get<1>(rowcol);
+  }
+
+  return 0;
+}
+
+bool OthelloState::CanCapture(Player player, int move) const {
+  if (board_[move] != CellState::kEmpty) return false;
+
+  std::tuple<int, int> row_col = XYFromCode(move);
+  int row = std::get<0>(row_col);
+  int col = std::get<1>(row_col);
+
+  bool valid = false;
+  for (int direction = Direction::kUp; direction < Direction::kLast; direction++) {
+    if (CountSteps(player, row, col, static_cast<Direction>(direction)) != 0) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void OthelloState::Capture(Player player, int row, int col, Direction dir, int steps) {
+  std::tuple<int, int> rowcol = GetNext(row, col, dir);
+  row = std::get<0>(rowcol);
+  col = std::get<1>(rowcol);
+  CellState cell = PlayerToState(player);
+
+  for (int step = 0; step < steps; step++) {
+    if (BoardAt(row, col) == CellState::kEmpty || BoardAt(row, col) == cell) {
+      SpielFatalError(absl::StrCat("Cannot capture cell (", row, ", ", col, ")"));
+    }
+
+    board_[row * kNumCols + col] = cell;
+    
+    rowcol = GetNext(row, col, dir);
+    row = std::get<0>(rowcol);
+    col = std::get<1>(rowcol);
+  }
+}
+
+int OthelloState::DiskCount(Player player) const {
+  int count = 0;
+  CellState cell = PlayerToState(player);
+  for (int i = 0; i < kNumCells; i++) {
+    if (board_[i] == cell) count++;
+  }
+
+  return count;
+}
+
+bool OthelloState::NoValidActions() const {
+  return (LegalRegularActions(Player(0)).empty() & LegalRegularActions(Player(1)).empty());
+}
+
+std::tuple<int, int> OthelloState::XYFromCode(int move) const {
+  if (move >= kNumCells || move < 0) {
+    SpielFatalError(absl::StrCat("Move too large: ", move));
+  }
+
+  return std::make_tuple(move / kNumCols, move % kNumRows);
+}
+
+CellState PlayerToState(Player player) {
+  switch (player) {
+    case 0:
+      return CellState::kBlack;
+    case 1:
+      return CellState::kWhite;
+    default:
+      SpielFatalError(absl::StrCat("Invalid player id ", player));
+      return CellState::kEmpty;
+  }
+}
+
+std::string StateToString(CellState state) {
+  switch (state) {
+    case CellState::kEmpty:
+      return ".";
+    case CellState::kWhite:
+      return "o";
+    case CellState::kBlack:
+      return "x";
+    default:
+      SpielFatalError("Unknown state.");
+  }
+}
+
+bool OthelloState::ValidAction(Player player, int move) const {
+  return (board_[move] == CellState::kEmpty && CanCapture(player, move));
+}
+
+void OthelloState::DoApplyAction(Action move) {
+  if (move == passMove) {  // pass
+    current_player_ = 1 - current_player_;
+    num_moves_ += 1;
+    return;
+  }
+
+  if (!ValidAction(current_player_, move)) {
+    SpielFatalError(absl::StrCat("Invalid move ", move));
+  }
+
+  std::tuple<int, int> rowcol = XYFromCode(move);
+  CellState cell = PlayerToState(current_player_);
+
+  int row = std::get<0>(rowcol);
+  int col = std::get<1>(rowcol);
+  board_[row * kNumCols + col] = cell;
```

Comment:
This x,y to move is done in a few places. Is it worth pulling into a function?

Reply:
Added the function RowColToMove(row, col) and RowColFromMove(move).